text/0000-must-use-output.md

```diff
@@ -0,0 +1,123 @@
+- Feature Name: `must-use-output`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow adding `#[must_use_output]` attribute on function paramters that triggers a warning if the function was called with a value that is later not used (only dropped). Also add the attribute to relevant library functions such as `Vec::push`.
+
+# Motivation
+[motivation]: #motivation
+
+Today, if you write this code in Rust:
+
+```rust
+fn main() {
+    let mut x = 0;
+    x += 42;
+}
+```
+
+you'll get a warning about unused write. It's likely that either you forgot to use the value of `x` or you wrote a needless operation that has no meaningful effect on the program.
+
+However this code:
+
+```rust
+fn main() {
+    let mut vec = Vec::new();
+    vec.push(42);
+}
+```
+
+has a similar problem and yet it doesn't trigger a warning. This can be even dangerous in some situations. For instance, if the `Vec` is collecting multiple errors (to report messages similar to `rustc`) and the code is supposed to check if the `Vec` is empty before proceeding.
+
+In some cases there are even multiple arguments that should be accessed after a call to a function. For instance `core::mem::swap` is almost always useless if both arguments are not accessed - if none of them are it only changes the drop order which doesn't matter in the vast majority of the code, and if only one of them is it's better to use a simple assignment rather than swap.
+
+Note also that some functions may be useless to call even when they take in truly immutable (`Freeze`) references. For instance `clone` is wasteful if the original is not accessed.
```

Comment:
```suggestion
```

This is true, but I don't think it's on-topic for *this* RFC, and addressing it would need some additional care to avoid the annoying ripple-effect problem of "don't write the clone on the *last* use". Can we drop it from this RFC?

Reply:
How about we mention it but don't propose to mark it just yet? Maybe some library has a use for it in some other context and I don't think adding code to ban `Freeze` references is worth it.

Reply:
@Kixunil I'm not suggesting adding code to *ban* it, just that it's something different and may not want to use the same handling. Could you move it to one of the later notes sections (e.g. future work), then, along with the note that this isn't proposing to mark such values?

Reply:
I'm confused. If we allow it right away and there's a clear use case with `clone`-like functions, which people could write how is it a "future possibility"? Note that while I wrote `clone` in the list below, I can simply remove it from the list and reword this to:

> Note also that some functions may be useless to call even when they take in truly immutable (`Freeze`) references. For instance `clone` is wasteful if the original is not accessed, however this RFC is **not** proposing adding the attribute to `core`'s `Clone::clone` just yet.

And put "Add the attribute to `clone`" in "Future possibilities"